File: crates/ox_content_renderer/tests/spec_support/spec_txt.rs
//! Loader for CommonMark-style `spec.txt` fixtures.
//!
//! The spec format stores each conformance case as a fenced block:
//!
//! ```text
//! `````````````````````````````````` example
//! markdown input
//! .
//! expected html
//! ``````````````````````````````````
//! ```
//!
//! Tabs are encoded as `→` inside examples so they survive editors that
//! expand tabs; the loader converts them back before returning.

/// One conformance example extracted from a spec fixture.
pub struct SpecExample {
    /// 1-based example number, matching the official spec numbering.
    pub number: usize,
    /// Title of the closest enclosing ATX section heading.
    pub section: String,
    /// Markdown input fed to the parser.
    pub markdown: String,
    /// Expected HTML output as printed in the spec.
    pub html: String,
}

const FENCE: &str = "````````````````````````````````";

/// Parses every example block out of a spec fixture.
pub fn parse_spec(text: &str) -> Vec<SpecExample> {
    let mut examples = Vec::new();
    let mut section = String::new();
    let mut lines = text.lines();

    while let Some(line) = lines.next() {
        if let Some(heading) = line.strip_prefix('#') {
            section = heading.trim_start_matches('#').trim().to_string();
            continue;
        }

        let is_example_fence = line
            .strip_prefix(FENCE)
            .is_some_and(|rest| rest.trim() == "example" || rest.trim().starts_with("example "));
        if !is_example_fence {
            continue;
        }

        let mut markdown = String::new();
        let mut html = String::new();
        let mut in_html = false;
        for body_line in lines.by_ref() {
            if body_line.starts_with(FENCE) {
                break;
            }
            if !in_html && body_line == "." {
                in_html = true;
                continue;
            }
            let target = if in_html { &mut html } else { &mut markdown };
            target.push_str(&body_line.replace('→', "\t"));
            target.push('\n');
        }

        examples.push(SpecExample {
            number: examples.len() + 1,
            section: section.clone(),
            markdown,
            html,
        });
    }

    examples
}

#[test]
fn parses_examples_and_sections() {
    let spec = format!(
        "# One\n\n{FENCE} example\na→b\n.\n<p>a\tb</p>\n{FENCE}\n\n\
         ## Two\n\n{FENCE} example table\n| a |\n.\n<table></table>\n{FENCE}\n"
    );
    let examples = parse_spec(&spec);
    assert_eq!(examples.len(), 2);
    assert_eq!(examples[0].number, 1);
    assert_eq!(examples[0].section, "One");
    assert_eq!(examples[0].markdown, "a\tb\n");
    assert_eq!(examples[0].html, "<p>a\tb</p>\n");
    assert_eq!(examples[1].number, 2);
    assert_eq!(examples[1].section, "Two");
    assert_eq!(examples[1].markdown, "| a |\n");
}

File: benchmarks/commonmark-conformance/spec-txt.mjs
/**
 * Loader for CommonMark-style `spec.txt` fixtures.
 *
 * Mirrors `crates/ox_content_renderer/tests/spec_support/spec_txt.rs` so the JS
 * conformance runner and the Rust conformance suite read the exact same
 * examples out of the exact same vendored spec file.
 *
 * The spec stores each case as a fenced block:
 *
 *     ```````````````````````````````` example
 *     markdown input
 *     .
 *     expected html
 *     ````````````````````````````````
 *
 * Tabs are encoded as `→` so they survive editors that expand tabs; they are
 * converted back here.
 */

const FENCE = "`".repeat(32);

/**
 * @typedef {Object} SpecExample
 * @property {number} number 1-based example number, matching official numbering
 * @property {string} section Title of the closest enclosing ATX heading
 * @property {string} markdown Markdown input fed to the parser
 * @property {string} html Expected HTML output as printed in the spec
 */

/**
 * @param {string} text
 * @returns {SpecExample[]}
 */
export function parseSpec(text) {
  const examples = [];
  const lines = text.split("\n");
  let section = "";

  for (let index = 0; index < lines.length; index++) {
    const line = lines[index];

    if (line.startsWith("#")) {
      section = line.replace(/^#+/, "").trim();
      continue;
    }

    if (!line.startsWith(FENCE)) continue;
    const rest = line.slice(FENCE.length).trim();
    if (rest !== "example" && !rest.startsWith("example ")) continue;

    let markdown = "";
    let html = "";
    let inHtml = false;

    for (index++; index < lines.length; index++) {
      const bodyLine = lines[index];
      if (bodyLine.startsWith(FENCE)) break;
      if (!inHtml && bodyLine === ".") {
        inHtml = true;
        continue;
      }
      const decoded = bodyLine.replaceAll("→", "\t") + "\n";
      if (inHtml) html += decoded;
      else markdown += decoded;
    }

    examples.push({ number: examples.length + 1, section, markdown, html });
  }

  return examples;
}
